lgc/patch/LowerDebugPrintf.cpp

```diff
@@ -0,0 +1,260 @@
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  LowerDebugPrintf.cpp
+ * @brief LLPC source file: contains implementation of class lgc::LowerDebugPrintf.
+ ***********************************************************************************************************************
+ */
+#include "lgc/patch/LowerDebugPrintf.h"
+#include "lgc/patch/Patch.h"
+#include "lgc/state/PalMetadata.h"
+#include "lgc/state/PipelineState.h"
+#include "llvm/ADT/Hashing.h"
+#include "llvm/BinaryFormat/MsgPackDocument.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "lower-debug-printf"
+
+using namespace llvm;
+using namespace lgc;
+
+namespace lgc {
+
+// =====================================================================================================================
+// Executes this LLVM patching pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+// @returns : The preserved analyses (The analyses that are still valid after this pass)
+PreservedAnalyses LowerDebugPrintf::run(Module &module, ModuleAnalysisManager &analysisManager) {
+  LLVM_DEBUG(dbgs() << "Run the pass Lower-debug-printf\n");
+  PipelineState *pipelineState = analysisManager.getResult<PipelineStateWrapper>(module).getPipelineState();
+  Patch::init(&module);
+  m_pipelineState = pipelineState;
+  SmallVector<CallInst *> callees;
+  BuilderBase builder(*m_context);
+  for (auto &func : module) {
+    auto name = func.getName();
+    if (name.startswith(lgcName::LowerDebugPrintf)) {
+      for (auto user : func.users()) {
+        if (CallInst *callInst = dyn_cast<CallInst>(user)) {
+          Value *resultVal = builder.getInt64(0);
+          if (!isa<PoisonValue>(callInst->getArgOperand(0))) {
+            builder.SetInsertPoint(callInst);
+            resultVal = createDebugPrintf(callInst->getArgOperand(0), callInst->getArgOperand(1),
+                                          make_range(callInst->arg_begin() + 2, callInst->arg_end()), builder);
+          }
+          callInst->replaceAllUsesWith(resultVal);
+          callees.push_back(callInst);
+        }
+      }
+    }
+  }
+
+  for (auto callInst : callees) {
+    callInst->eraseFromParent();
+  }
+
+  if (m_elfInfos.size() == 0)
+    return PreservedAnalyses::none();
```

Comment:
Please just use `m_elfInfos.empty()`.

Also, the logic here still looks wrong to me:

* If m_elfInfos is empty, then this pass made no changes. So, we should return `PreservedAnalyses::all()`: all analyses were preserved.
* If m_elfInfos is not empty, then this pass did make changes. This means some analyses need to be discarded. This pass doesn't change the CFG, so it should return `PreservedAnalyses::allInSet<CFGAnalyses>()`.